subprojects/core/src/main/java/org/gradle/internal/build/PlannedNodeGraph.java

```diff
@@ -0,0 +1,235 @@
+            }
+        }
+
+        return newPlannedNodes;
+    }
+
+    public enum DetailLevel {
+        LEVEL1_TASKS(1, NodeType.TASK),
+        LEVEL2_TRANSFORM_STEPS(2, NodeType.TASK, NodeType.TRANSFORM_STEP);
+
+        private final int level;
+        private final Set<NodeType> nodeTypes;
+
+        DetailLevel(int level, NodeType... nodeTypes) {
+            this.level = level;
+            this.nodeTypes = EnumSet.copyOf(Arrays.asList(nodeTypes));
+        }
+
+        public int getLevel() {
+            return level;
+        }
+
+        public boolean contains(NodeType nodeType) {
+            return nodeTypes.contains(nodeType);
+        }
+
+        public static DetailLevel from(Set<NodeType> nodeTypes) {
+            for (DetailLevel detailLevel : values()) {
+                if (detailLevel.nodeTypes.equals(nodeTypes)) {
+                    return detailLevel;
+                }
+            }
+
+            throw new IllegalStateException("Unknown detail level for node types: " + nodeTypes);
+        }
+    }
+
+    /**
+     * {@link #collectNodes(Collection) Collects} and converts nodes to planned nodes
+     * resolving their dependencies with the highest available {@link DetailLevel detail level}.
+     */
+    public static class Collector {
+
+        private final ToPlannedNodeConverterRegistry converterRegistry;
+        private final DetailLevel detailLevel;
+
+        private final List<PlannedNodeInternal> plannedNodes = new ArrayList<>();
+        // This cache is necessary, because PlannedNodeGraph relies on node identities being reference-equal
+        private final Map<Node, NodeIdentity> nodeIdentityCache = new IdentityHashMap<>();
+
+        public Collector(ToPlannedNodeConverterRegistry converterRegistry) {
+            this.converterRegistry = converterRegistry;
+            this.detailLevel = DetailLevel.from(converterRegistry.getConvertedNodeTypes());
+        }
+
+        public DetailLevel getDetailLevel() {
+            return detailLevel;
+        }
+
+        public void collectNodes(Collection<Node> nodes) {
+            for (Node node : nodes) {
+                ToPlannedNodeConverter converter = converterRegistry.getConverter(node);
+                if (converter != null && converter.isInSamePlan(node)) {
+                    List<? extends NodeIdentity> nodeDependencies = findNodeDependencies(node);
+                    PlannedNodeInternal plannedNode = converter.convert(node, nodeDependencies, () -> findTaskDependencies(node));
+                    plannedNodes.add(plannedNode);
+                }
+            }
+        }
+
+        public PlannedNodeGraph getGraph() {
+            return new PlannedNodeGraph(detailLevel, plannedNodes);
+        }
+
+        private List<? extends NodeIdentity> findNodeDependencies(Node node) {
+            return findDependencies(node, Node::getDependencySuccessors, it -> true);
```

Comment:
❔ IIUC, we don't have any ordering guarantees here for the dependencies, nor did we have them before. Is that correct?